Index: src/storage/DatabaseAdapter.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/storage/DatabaseAdapter.php	(revision 871b7d27896d9860f92a50f021e9ba054f6827f1)
+++ src/storage/DatabaseAdapter.php	(revision ba02e72904627cf3713e9857bc20166744b31f8e)
@@ -291,8 +291,9 @@
                             }
                         }
 
-                        if ($db["num_rows"] < $this::MAX_NUMROWS) {
-                            if ($this->rawdata || $db["num_rows"] > $this::MAX_RESULTS) {
+                        $count_recordset                            = count($db["recordset"]);
+                        if (!empty($query["limit"]) || $count_recordset < $this::MAX_NUMROWS) {
+                            if ($this->rawdata || $count_recordset > $this::MAX_RESULTS) {
                                 $res["rawdata"]                     = $db["recordset"];
                             } else {
                                 $key                                = $this->getFieldAlias($query["key"]);
