Index: phpformsframework/libs/src/storage/OrmModel.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- phpformsframework/libs/src/storage/OrmModel.php	(date 1568203068844)
+++ phpformsframework/libs/src/storage/OrmModel.php	(date 1568203068844)
@@ -207,6 +207,16 @@
 
         return Orm::readRawData($where, $fields, $sort, $limit, $this);
     }
+
+    /**
+     * @param array $data
+     * @return array|bool|null
+     */
+    public function insertUnique($data)
+    {
+        return Orm::insertUnique($data, $this);
+    }
+
     /**
          * @param array $data
          * @return array|bool|null
